mobly/controllers/android_device.py

```diff
@@ -466,6 +477,74 @@ def debug_tag(self, tag):
         self._debug_tag = tag
         self.log.extra['tag'] = tag
 
+    @property
+    def has_active_service(self):
+        """True if any service is running on the device.
+
+        A service can be a snippet or logcat collection.
+        """
+        return any(
+            [self._snippet_clients, self._adb_logcat_process, self.sl4a])
+
+    @property
+    def log_path(self):
+        """A string that is the path for all logs collected from this device.
+
+        This by design should not be changed during the test.
+        """
+        return self._log_path
+
+    @log_path.setter
+    def log_path(self, new_path):
+        if self.has_active_service:
+            raise DeviceError(
+                self,
+                'Cannot change `log_path` when there is service running.')
+        old_path = self._log_path
+        utils.create_dir(new_path)
+        if os.path.exists(old_path):
+            distutils.dir_util.copy_tree(self._log_path, new_path)
+            shutil.rmtree(self._log_path, ignore_errors=True)
+        self._log_path = new_path
+
+    @property
+    def serial(self):
+        """The serial number used to identify a device.
+
+        This is essentially the value used for adb's `-s` arg, which means it
+        can be a network address or USB bus number.
+        """
+        return self._serial
+
+    def update_serial(self, new_serial):
```

Comment:
In function start_adb_logcat of android_device.py 
Please change fname to -
f_name = 'adblog,%s,%s.txt' % (self.model, self.debug_tag)
As otherwise after changing serial of remote device, this would result in another logfile with new serial.

For remote devices, users can setup custom device name with debug_tag, before starting logcat service.

Reply:
If the serial changed in the middle of the test, it seems good to have the adb logs split between before and after the change?

Reply:
the problem is, debug_tag can change at any time without guarding.
I don't want file names to be dependent on that.
If your device changes serial in the middle of a test, that's kind of what you have to suffer atm.

The alternative is to force all existing users to conform to the same approach described in this PR, which is more costly.

Reply:
Acknowledged.